test/Mono.Linker.Tests.Cases/DataFlow/NullableAnnotations.cs

```diff
@@ -0,0 +1,289 @@
+			RequirePublicFieldsOnGenericParam<Nullable<StructWithFieldsReferencedThroughDamOnNullable>> ();
+			m<StructWithFieldsReferencedThroughDamOnNullable> ();
+
+			GenericClassTypeParamRequiresPublicFields<GenericClassTypeParam>.RunTests ();
+			GenericClassUnderlyingTypeParamRequiresPublicFields<Nullable<GenericClassTypeParam>>.RunTests ();
+
+			TestGetUnderlyingTypeOnClasses ();
+			TestGetUnderlyingTypeOnStructs ();
+			TestGetUnderlyingTypeOnNullableStructs ();
+			TestGetUnderlyingTypeOfCreatedNullableOnStructs ();
+		}
+
+		[Kept]
+		[ExpectedWarning ("IL2026", "message")]
+		static void RequireMethodWithRUC ()
+		{
+			var T = typeof (Nullable<TestStructWithRucMethod>);
+			var uT = Nullable.GetUnderlyingType (T);
+			uT.RequiresAll ();
+		}
+
+		[Kept]
+		static void UnderlyingTypeOfAnnotatedGenericParameterRequiresPublicProperties<[KeptAttributeAttribute (typeof (DAM))][DAM (DAMT.PublicProperties)] TNullable> ()
+		{
+			(Nullable.GetUnderlyingType (typeof (TNullable))).RequiresPublicProperties ();
+		}
+
+		[Kept]
+		static void UnderlyingTypeOfAnnotatedParameterRequiresPublicProperties ([KeptAttributeAttribute (typeof (DAM))][DAM (DAMT.PublicProperties)] Type tNullable)
+		{
+			(Nullable.GetUnderlyingType (tNullable)).RequiresPublicProperties ();
+		}
+
+		[Kept]
+		[ExpectedWarning ("IL2067")]
+		static void UnderlyingTypeOfUnannotatedParameterRequiresPublicProperties (Type tNullable)
+		{
+			(Nullable.GetUnderlyingType (tNullable)).RequiresPublicProperties ();
+		}
+
+		[Kept]
+		[ExpectedWarning ("IL2087")]
+		static void UnderlyingTypeOfUnannotatedGenericParameterRequiresProperties<TNullable> ()
+		{
+			(Nullable.GetUnderlyingType (typeof (TNullable))).RequiresPublicProperties ();
+		}
+
+		[Kept]
+		static void NullableOfAnnotatedGenericParameterRequiresPublicProperties<[KeptAttributeAttribute (typeof (DAM))][DAM (DAMT.PublicProperties)] T> () where T : struct
+		{
+			(Nullable.GetUnderlyingType (typeof (Nullable<T>))).RequiresPublicProperties ();
+		}
+
+		[Kept]
+		[ExpectedWarning ("IL2087")]
+		static void NullableOfUnannotatedGenericParameterRequiresPublicProperties<T> () where T : struct
+		{
+			(Nullable.GetUnderlyingType (typeof (Nullable<T>))).RequiresPublicProperties ();
+		}
+
+		[Kept]
+		static void NullableOfAnnotatedParameterRequiresPublicProperties ([KeptAttributeAttribute (typeof (DAM))][DAM (DAMT.PublicProperties)] Type t)
+		{
+			(Nullable.GetUnderlyingType (typeof (Nullable<>).MakeGenericType (t))).RequiresPublicProperties ();
+		}
+
+		[Kept]
+		static void TestGetUnderlyingTypeOnClasses ()
+		{
+			UnderlyingTypeOfAnnotatedParameterRequiresPublicProperties (typeof (TestClass));
+			UnderlyingTypeOfAnnotatedGenericParameterRequiresPublicProperties<TestClass> ();
+			UnderlyingTypeOfUnannotatedParameterRequiresPublicProperties (typeof (TestClass));
+			UnderlyingTypeOfUnannotatedGenericParameterRequiresProperties<TestClass> ();
+		}
+
+		[Kept]
+		static void TestGetUnderlyingTypeOnStructs ()
```

Comment:
I don't see the value of having both class and struct tests for this. Basically what this does is validating that applying PublicProperties annotation to a class or struct works correctly. While a valid test, it has nothing to do with nullable. The fact that the method it calls does some nullable magic inside, is not seen at this callsite.